scripts/inference/benchmarking/README.md

```diff
@@ -14,4 +18,118 @@ To run the benchmarking script, use the following command:
+
+##### KV cache
+
+One of the important elements of LLM inference is a Key-Value (KV) cache. As part of the prefill stage, the key and value tensors are computed and cached for each token in the prompt across the model layers. This is helpful because the initial prompt does not change during inference, which allows the model layers to simply re-use the cached values, rather than spending compute on re-computing the tokens for the context. \
+To set up the KV cache for benchmarking, set `use_cache: true` in the `.yaml` file.
+
+
+### Results
+Now that we have the necessary background to reason about inference, we can move onto the fun stuff, ie. real-world benchmark results.
+
+TODO: we will be releasing the configs used to reproduce these latency results soon!
+### I want to run generation with MPT-[X], how long does it take to process different size inputs?
+
+#### Setup
+We use a single A100 80GB for inference, running with precision `bf16` and the `torch` implementation of attention. We include the highest possible batch size that is able to produce 1000 output tokens without running out of GPU memory.
+
+Here we show how latency changes for a given input prompt length, while varying batch size and output length.
+This gives a rule of thumb of how fast you can expect MPT to be based on different generation parameters.
+#### TL;DR
+Hardware Setup:
+- 1 x NVIDIA A100 80GB
+
+
+Benchmark Setup:
+- Input Length: [128, 2048]
+- Output Length: [1, 10, 100, 1000]
+#### Short Inputs (128 input tokens) on MPT-7B
+![assets](assets/Latency-for-MPT-7B,-n_input_tok=128.svg)
+#### Long Inputs (2048 input tokens) on MPT-7B
+![assets](assets/Latency-for-MPT-7B,-n_input_tok=2048.svg)
+#### Short Inputs (128 input tokens) on MPT-30B
+![assets](assets/Latency-for-MPT-30B,-n_input_tok=128.svg)
+#### Long Inputs (2048 input tokens) on MPT-30B
+![assets](assets/Latency-for-MPT-30B,-n_input_tok=2048.svg)
+
+Our real-world results match the theory! The latency grows nearly linearly with the number of output tokens, which is the _decode_ stage time. For large batch sizes and output lengths, the latency follows more of a quadratic, which is the attention operation overhead kicking in.
+
+For longer input lengths and batch sizes, the _prefill_ stage starts to become more important, given that the model has to process a lot of input tokens in the forward pass.
+Despite the _prefill_ stage being highly efficient, the model still needs to perform a lot of computation in one forward pass, which results in the higher latency when increasing batch size and input length.
+
+In general, latency tends to be dominated by long sequences rather than larger batch sizes, so whenever possible, it's valuable to increase batch sizes as this allows for processing a lot more inputs without (much) additional cost.
```

Comment:
Attention has quadratic complexity in the seq dim not batch size.
Should the above exclude comments on batch size?

Reply:
resolved offline